src/main/kotlin/org/dif/peerdid/PeerDIDUtils.kt

```diff
@@ -0,0 +1,149 @@
+        if (PublicKeyTypeAuthentication.values().any { value -> value.name == decodedEncnumbasises[i]["type"] })
+            authentication.add(decodedEncnumbasises[i])
+        else if (PublicKeyTypeAgreement.values().any { value -> value.name == decodedEncnumbasises[i]["type"] })
+            keyAgreement.add(decodedEncnumbasises[i])
+        else
+            throw IllegalArgumentException("Invalid key type of: ${keysList[i]}")
+    }
+
+    val gson = GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create()
+    val diddoc = mapOf(
+        "id" to peerDID,
+        "authentication" to gson.toJsonTree(authentication),
+        "keyAgreement" to gson.toJsonTree(keyAgreement),
+        "service" to gson.toJsonTree(decodedService)
+    )
+    return gson.toJson(diddoc)
+}
+
+private fun isInEncodingTypes(encodingAlgorithm: Char): Boolean {
+    return EncodingType.values().any { type -> type.type == encodingAlgorithm }
+}
+
+private fun decodeService(encodedService: JSON, peerDID: PeerDID): Map<String, String> {
+    val decodedService = Base64.decodeBase64(encodedService).decodeToString()
+    val gson = GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create()
+    val serviceMap = gson.fromJson(decodedService, HashMap::class.java)
+    val serviceType = serviceMap.remove("t").toString().replace("dm", "didcommmessaging")
+
+    return mapOf(
+        "id" to peerDID.plus('#').plus(serviceType),
+        "type" to serviceType,
+        "serviceEndpoint" to serviceMap.remove("s").toString(),
+        "routingKeys" to serviceMap.remove("r").toString()
+    )
+}
+
+private fun decodeEncnumbasis(encnumbasis: String, peerDID: PeerDID): Map<String, String> {
+    val decodedEncnumbasis = Base58.decode(encnumbasis)
+    val codec = getCodec(decodedEncnumbasis)
+    val decodedEncnumbasisWithoutPrefix = removePrefix(decodedEncnumbasis)
+    val publicKey = Base58.encode(decodedEncnumbasisWithoutPrefix.toByteArray())
+    return mapOf(
+        "id" to peerDID.plus('#').plus(encnumbasis),
+        "type" to codec,
+        "controller" to peerDID,
+        "publicKeyBase58" to publicKey
+    )
+}
+
+private fun getCodec(data: ByteArray): String {
+    val prefix = extractPrefix(data)
+    PublicKeyTypeAgreement.values().forEach { type -> if (type.prefix() == prefix) return type.name }
+    PublicKeyTypeAuthentication.values().forEach { type -> if (type.prefix() == prefix) return type.name }
+    throw IllegalArgumentException("Prefix $prefix not supported")
+}
+
+private fun extractPrefix(data: ByteArray): Int {
+    return VarInt.readVarint(ByteBuffer.wrap(data))
+}
+
+private fun removePrefix(data: ByteArray): List<Byte> {
+    val prefixInt = extractPrefix(data)
+    val byteBuffer = ByteBuffer.allocate(2)
+    VarInt.writeVarInt(prefixInt, byteBuffer)
+    return data.drop(byteBuffer.array().size)
+}
+
+internal fun createEncnumbasis(key: PublicKey<out PublicKeyType>): String {
+    val decodedKey = Base58.decode(key.encodedValue)
```

Comment:
What if the key value is not a correct base58? I think `IllegalArgumentException` should be raised.